### DISCUSSION TITLE
Disable billing fields for free checkout (block-based)

### DISCUSSION THREAD

**Discussion by andkit01** [woocommerce/woocommerce#45568] [Checkout Flow]
When the checkout total is zero, is there a way to disable the billing fields, using the block-based checkout. The old filter hooks like woocommerce_checkout_fields don't seem to work anymore. Thanks!

**Comment by senadir** [comment]
Hey! This is not possible yet, but we have it on our radar.

One question, if you're to disable billing, what would you collect exactly? just the email? If that so, this can be done with a PHP + CSS code snippet.

**Reply by senadir** [reply]
This is how it can be done:

Disable Phone, Company, and Address 2 from the editor:
<img width="287" alt="image" src="https://github.com/woocommerce/woocommerce/assets/6165348/1b66d55d-0cc4-41f9-9716-bdddc6137364">

Base your customer default address to your store address:
<img width="696" alt="image" src="https://github.com/woocommerce/woocommerce/assets/6165348/4cc95f77-f0fc-4fa2-96b6-73807055ea2d">

Disable all fields for US if US is your store country for example:
```php

add_filter(
	'woocommerce_get_country_locale',
	function ( $locale ) {
		$cart = WC()->cart;
		// Only remove fields if we have a cart.
		if ( ! $cart ) {
			return $locale;
		}

		// If cart requires shipping, then it's likely paid and isn't calculated yet.
		if ( $cart->needs_shipping() ) {
			return $locale;
		}
		// Only remove fields if the cart total is 0.
		if ( $cart->get_total( 'edit' ) > 0 ) {
			return $locale;
		}

		$locale['US'] = wp_parse_args(
			array(
				'state'      => array(
					'hidden'   => 'true',
					'required' => 'false',
				),
				'city'       => array(
					'hidden'   => 'true',
					'required' => 'false',
				),
				'postcode'   => array(
					'hidden'   => 'true',
					'required' => 'false',
				),
				'first_name' => array(
					'hidden'   => 'true',
					'required' => 'false',
				),
				'last_name'  => array(
					'hidden'   => 'true',
					'required' => 'false',
				),
				'address_1'  => array(
					'hidden'   => 'true',
					'required' => 'false',
				),
			),
			$locale['US'],
		);
		return $locale;
	}
);
```

You will get a Checkout like this:
<img width="1146" alt="image" src="https://github.com/woocommerce/woocommerce/assets/6165348/33e03608-38b7-4151-8927-3cb710a2360b">


You can proceed by hiding the billing step if totals are less than 0:

```php

add_action(
	'wp_footer',
	function () {
		if ( is_checkout() ) {
			?>
			<script>
				// Add an event to pag load
				window.addEventListener("load", function () {
					const cartTotals = wp.data
						.select("wc/store/cart")
						?.getCartTotals().total_price;

					if (cartTotals === undefined || cartTotals > 0) {
						return;
					}

					const billingStep = document.querySelector(
						".wc-block-checkout__billing-fields"
					);

					if (!billingStep) {
						return;
					}
					// Add hidden styles to the billingStep
					billingStep.style.display = "none";
				});

			</script>
			<?php
		}
	}
);
```

**Reply by senadir** [reply]
@helgatheviking this might be a good basis for your plugin?

**Comment by andkit01** [comment]
Hi Seghir,
Thank you very much for the spot-on reply and code, it is very helpful and
much appreciated!!

"...Disable billing, what would you collect exactly? just the email?..."
Yes, the idea is for the Free orderable item, only the email (to register
the user) is collected, and no billing address or payment information is
needed. I would also like to disable the order confirmation page and email.
I ended up switching to the old checkout page, and was able to use all the
hooks to achieve this (remove the billing fields, disable the order
confirmation page and email for the Free checkout). For Paid checkout, the
normal flow applies.

It looks like the woocommerce_get_country_locale hook affects the
block-based checkout, even though unfortunately
the woocommerce_checkout_fields hook doesn't seem to work for the
block-based checkout. It would be nice if this hook would work for the
block-based checkout, not just the old checkout, but it seems that your
solution would work just as effectively. And I suppose setting the default
country to the US helps to avoid any field validation error, so we can just
hide it by css using the javascript. I see.

I think I'm good for now, thank you again for the help Seghir!


On Mon, Mar 18, 2024 at 10:39 AM Seghir Nadir ***@***.***>
wrote:

> This is how it can be done:
>
> Disable Phone, Company, and Address 2 from the editor:
> image.png (view on web)
> <https://github.com/woocommerce/woocommerce/assets/6165348/1b66d55d-0cc4-41f9-9716-bdddc6137364>
>
> Base your customer default address to your store address:
> image.png (view on web)
> <https://github.com/woocommerce/woocommerce/assets/6165348/4cc95f77-f0fc-4fa2-96b6-73807055ea2d>
>
> Disable all fields for US if US is your store country for example:
>
> add_filter(
> 	'woocommerce_get_country_locale',
> 	function ( $locale ) {
> 		$cart = WC()->cart;
> 		// Only remove fields if we have a cart.
> 		if ( ! $cart ) {
> 			return $locale;
> 		}
>
> 		// If cart requires shipping, then it's likely paid and isn't calculated yet.
> 		if ( $cart->needs_shipping() ) {
> 			return $locale;
> 		}
> 		// Only remove fields if the cart total is 0.
> 		if ( $cart->get_total( 'edit' ) > 0 ) {
> 			return $locale;
> 		}
>
> 		$locale['US'] = wp_parse_args(
> 			array(
> 				'state'      => array(
> 					'hidden'   => 'true',
> 					'required' => 'false',
> 				),
> 				'city'       => array(
> 					'hidden'   => 'true',
> 					'required' => 'false',
> 				),
> 				'postcode'   => array(
> 					'hidden'   => 'true',
> 					'required' => 'false',
> 				),
> 				'first_name' => array(
> 					'hidden'   => 'true',
> 					'required' => 'false',
> 				),
> 				'last_name'  => array(
> 					'hidden'   => 'true',
> 					'required' => 'false',
> 				),
> 				'address_1'  => array(
> 					'hidden'   => 'true',
> 					'required' => 'false',
> 				),
> 			),
> 			$locale['US'],
> 		);
> 		return $locale;
> 	}
> );
>
> You will get a Checkout like this:
> image.png (view on web)
> <https://github.com/woocommerce/woocommerce/assets/6165348/33e03608-38b7-4151-8927-3cb710a2360b>
>
> You can proceed by hiding the billing step if totals are less than 0:
>
> add_action(
> 	'wp_footer',
> 	function () {
> 		if ( is_checkout() ) {
> 			?>
> 			<script>
> 				// Add an event to pag load
> 				window.addEventListener("load", function () {
> 					const cartTotals = wp.data
> 						.select("wc/store/cart")
> 						?.getCartTotals().total_price;
>
> 					if (cartTotals === undefined || cartTotals > 0) {
> 						return;
> 					}
>
> 					const billingStep = document.querySelector(
> 						".wc-block-checkout__billing-fields"
> 					);
>
> 					if (!billingStep) {
> 						return;
> 					}
> 					// Add hidden styles to the billingStep
> 					billingStep.style.display = "none";
> 				});
>
> 			</script>
> 			<?php
> 		}
> 	}
> );
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/woocommerce/woocommerce/discussions/45568#discussioncomment-8828981>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJ7F6R7PDQV7XS5ADB5SGA3YY34ATAVCNFSM6AAAAABEVJZQB6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DQMRYHE4DC>
> .
> You are receiving this because you authored the thread.
>

**Comment by lamanrasmi** [comment]
Hi @andkit01 , i tried your snippets but it did not work with the latest Woo. none of the fields removed. Is there any other solutions for the block based checkout?

**Reply by sandraleung** [reply]
I got it working with the modified code below.
Just change the fields as needed
```
add_filter(
	'woocommerce_get_country_locale',
	function ( $locale ) {
		$locale['HK'] = wp_parse_args(
			array(
				'city'       => array(
					'hidden'   => 'true',
					'required' => 'false',
				),
				'postcode'   => array(
					'hidden'   => 'true',
					'required' => 'false',
				),

			),
			$locale['HK'],
		);
		return $locale;
	}
);
```

**Comment by senadir** [comment]
that one of for block based checkout and i tested it with it. did you see any errors?

**Comment by bridgeport** [comment]
While the above posted code does work to remove the specified billing address fields from view, checkout fails when the form is submitted due to backend validation still expecting those fields.

It fails with the following message:

> There was a problem with the provided billing address: Postcode / ZIP is required, State / County is required, Street address is required, Town / City is required

It appears to take place in the `validate_required_additional_fields()` function within:

> /wp-content/plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php

Is there any way to supplement the frontend removals by suppressing the backend validation for those fields?

Thanks.

**Comment by marcosgomesneto** [comment]
To fix the validation, you need to adjust the code as there is an error. Currently, the field properties are being set as the strings "true" and "false", but they should be replaced with the boolean values **true** and **false** without quotes. This correction will make the field validation work properly and ensure that hidden fields are ignored during the checkout process.

If you are trying to hide the fields, as the above code does not work, try hiding the fields from the default locale:

```php
add_filter( 'woocommerce_get_country_locale_base', function( $locale ) {
		$cart = WC()->cart;

		if ( ! $cart ) {
			return $locale;
		}

		if ( $cart->needs_shipping() ) {
			return $locale;
		}

		return wp_parse_args(
			[ 
				'state' => [ 
					'hidden' => true,
					'required' => false,
				],
				'city' => [ 
					'hidden' => true,
					'required' => false,
				],
				'postcode' => [ 
					'hidden' => true,
					'required' => false,
				],
				'address_1' => [ 
					'hidden' => true,
					'required' => false,
				],
			],
			$locale
		);

	}

		);
```

**Comment by bridgeport** [comment]
Fantastic. Thanks @marcosgomesneto that fixed it.

**Comment by alif13579** [comment]
Thanks for the loop. In block based checkout page, there are some fields unnecessary but required. So i already found this loop is helpful but want some solutions like email field, country field, These need to set non required and hidden and also phone field need to required and type to number. 

I can't do that using multiple keys. Please help @marcosgomesneto

**Reply by senadir** [reply]
Phone type is phone (not number). You can mark the phone as optional from the editor directly.

For Email and country it's not currently possible to do that in a clean way but we will look into it.

**Comment by senadir** [comment]
This is currently the best available tutorial for disabling all fields

https://nadir.blog/2024/07/22/removing-fields-from-checkout-block/

**Reply by froger-me** [reply]
The reasons given under "Why we don’t encourage removing fields" seem quite invalid: if the payment processors need these fields, it should be their responsibility to enforce their presence, not the default behavior to prevent their alteration.  
There are ways to do that, with existing filters for fields when not using the block editor (example: save the fields early and add an attribute to signal they're necessary to subsequent filters, and reapplying the necessary fields late if they have been removed), instead of using an unrelated  `woocommerce_get_country_locale` filter.

Many QR-based payment methods do not need these fields for fraud detection (example: Alipay, WeChat Pay, which provide an extremely low amount of friction, and for this very reason made them extremely popular in China), and many businesses handle taxes outside of WooCommerce.  

Saying "This is intentional, we’re taking back control of how to render and handle fields" feels like a regression.

**Comment by astappiev** [comment]
The solutions described here don't work after an update to 9.8.2, was working fine on 9.7.1.

This one in particular 
https://github.com/woocommerce/woocommerce/blob/trunk/docs/cart-and-checkout-blocks/removing-checkout-fields.md

I needed to remove `'hidden' => true,` and hide the fields via CSS to make it working. Otherwise `Place order` button does nothing, no errors are shown, but the order process is blocked (probably by validation).
